compiler/daml-lf-reader/src/DA/Daml/LF/Reader.hs

```diff
@@ -111,13 +121,27 @@ readDalfManifest dar = do
         lookup attrName attrs
     missingAttr attrName = Left $ "No " <> BSUTF8.toString attrName <> " attribute in manifest."
 
-readDalfs :: Archive -> Either String Dalfs
+readDalfs :: Archive -> Either String (Dalfs BSL.ByteString)
 readDalfs dar = do
     DalfManifest{..} <- readDalfManifest dar
     mainDalf <- getEntry dar mainDalfPath
     dalfs <- mapM (getEntry dar) (delete mainDalfPath dalfPaths)
     pure $ Dalfs mainDalf dalfs
 
+extractNameAndPackageIdFromPath :: FilePath -> (T.Text, LF.PackageId)
+extractNameAndPackageIdFromPath = bimap (T.pack . intercalate "-") (LF.PackageId . T.pack) . fromJust . unsnoc . wordsBy (=='-') . takeBaseName
```

Comment:
TODO: Reconsider this safety...
The PackageId can be derived from the ArchivePayload
The unit-id requires the entire package to be read through proto, so we can read the metadata. Not ideal, perhaps theres some way to lazily parse?

Reply:
I've been looking into this for longer than I wanted but couldn't get anywhere; the main blocker is that `proto3-suite`/`proto3-wire` are strict by default because it reads in a `Parser` _monad_ so later results depend on earlier ones. 

One thing we could do is make a stripped down version of the proto files with just `ArchivePayload`, `Package`, `PackageMetadata` and the interned strings, plus the attending haskell files for that, and use that for this purpose, but that sounds like a lot of copying and pasting. 

We could also take the daml2/3 break to move the metadata up into an intermediate level, I don't think we've made any promises for the daml3 encoded dars yet?

Reply:
IThe proto format of LF2 is definitely not set in stone, but I'd double-check with @remyhaemmerle-da  whether that's worth it because it would be our first truly breaking change.